Hubspot contact and company association

pauloouriques opened the pull request:
## Change description

This PR makes it possible to associate a contact to a company and also adds all re required tests to cover that new feature.

## Checklists

### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

pauloouriques committed:
chore: contact > company association and tests.

evantahler commented on plugins/@grouparoo/hubspot/src/lib/export-contacts/connection.ts:
```diff
@@ -20,7 +20,14 @@ export const contactsDestinationConnection: PluginConnection = {
+      key: "companyKey",
+      displayName: "Company Key",
+      required: false,
+      description: "Which field in the Company uniquely represents it?",
```
If companyKey is optional, what will be used by default?  We might want to mention the email domain will be used if this is not set.

pauloouriques replied:
If `companyKey` is not set the relationship is just ignored.

evantahler replied:
But doesn't hubspot make companies autoamtically from the email domain?  Either way, 👍

bleonard replied:
This could say something like "Links a Contact to a company by something other than the Hubspot default (email domain)."

pauloouriques replied:
ahh! Now I got what you're saying @evantahler, thank you. Brian's text will do the trick.

bleonard commented on plugins/@grouparoo/hubspot/__tests__/.env.example:
```diff
@@ -14,3 +14,6 @@ HUBSPOT_API_KEY = "a31ef454-1835-4b88-8d3a-eefc0b688ded"
 # last_name                   (String)
 # text_field                  (String)
 # number_field                (Number)
+#The Hubspot account should have 2 companies with the following domains:
```
It would be nice if the tests added these if they are necessary.

bleonard replied:
I think it's ok to require a schema be set up but data should be added by the tests

bleonard commented on plugins/@grouparoo/hubspot/src/lib/export-contacts/destinationMappingOptions.ts:
```diff
@@ -8,9 +8,15 @@ import { connect } from "../connect";
+
+    if (destinationOptions?.companyKey) {
+      const type =
+        destinationOptions.companyKey === "phone" ? "phoneNumber" : "string";
```
seems possible it's an integer?

bleonard commented on plugins/@grouparoo/hubspot/src/lib/export-contacts/exportRecord.ts:
```diff
@@ -88,6 +97,9 @@ export const exportRecord: ExportRecordPluginMethod = async ({
       );
       const formattedDataFields = {};
       for (const key of Object.keys(payload)) {
+        if (companyMappingKey && key === companyMappingKey) {
```
something funny her about `old` Company keys. Like the setting got switched from `domain` to `company_id`
Maybe it should skip anything that starts with `Company.` 
Then the old one would get sent to the contact.

bleonard commented on plugins/@grouparoo/hubspot/src/lib/export-contacts/exportRecord.ts:
```diff
@@ -124,6 +136,22 @@ export const exportRecord: ExportRecordPluginMethod = async ({
+      if (companyKey) {
+        const oldCompanyKeyValue = oldRecordProperties[companyMappingKey];
+        const newCompanyKeyValue = newRecordProperties[companyMappingKey];
+        if (!contact) {
```
can this be returned from `createOrUpdateContact`

pauloouriques replied:
createOrUpdateContact returns undefined

pauloouriques committed:
chore: PR required adjusts.
 - include interger to destination options mapping type;
 - add test companies;
 - change destination option description;
 - skip anything that starts with Company.

pauloouriques committed:
chore: declare DestinationMappingOptionsResponseType type.
